crypto/keyutil/generate.go

```diff
@@ -0,0 +1,264 @@
+		}
+
+		pub, pk, err := ed25519.GenerateKey(r)
+		if err != nil {
+			return nil, nil, fmt.Errorf("unable to generate OKP key pair: %w", err)
+		}
+		return pub, pk, nil
+	default:
+	}
+
+	return nil, nil, errors.New("key type not supported")
+}
+
+// PublicKey extracts a public key from a private key.
+func PublicKey(priv any) (crypto.PublicKey, error) {
+	switch k := priv.(type) {
+	case *rsa.PrivateKey:
+		if k == nil {
+			return nil, errors.New("private key is nil")
+		}
+		return &k.PublicKey, nil
+	case *ecdsa.PrivateKey:
+		if k == nil {
+			return nil, errors.New("private key is nil")
+		}
+		return &k.PublicKey, nil
+	case ed25519.PrivateKey:
+		if k == nil {
+			return nil, errors.New("private key is nil")
+		}
+		return k.Public(), nil
+	case *ecdh.PrivateKey:
+		if k == nil {
+			return nil, errors.New("private key is nil")
+		}
+		return k.Public(), nil
+	case jose.JSONWebKey:
+		if k.IsPublic() {
+			return k.Key, nil
+		}
+		return k.Public().Key, nil
+	case *jose.JSONWebKey:
+		if k == nil {
+			return nil, errors.New("private key is nil")
+		}
+		if k.IsPublic() {
+			return k.Key, nil
+		}
+		return k.Public().Key, nil
+	case *rsa.PublicKey, *ecdsa.PublicKey, ed25519.PublicKey, *ecdh.PublicKey:
+		if k == nil {
+			return nil, errors.New("public key is nil")
+		}
+		return k, nil
+	default:
+		return nil, fmt.Errorf("unrecognized key type: %T", priv)
+	}
+}
+
+// ExtractKey returns the given public or private key or extracts the public key
+// if a x509.Certificate or x509.CertificateRequest is given.
+func ExtractKey(in any) (any, error) {
+	switch k := in.(type) {
+	case *rsa.PublicKey, *rsa.PrivateKey,
+		*ecdsa.PublicKey, *ecdsa.PrivateKey,
+		*ecdh.PublicKey, *ecdh.PrivateKey,
+		ed25519.PublicKey, ed25519.PrivateKey:
+		return in, nil
+	case []byte:
+		return in, nil
+	case *x509.Certificate:
+		return k.PublicKey, nil
+	case *x509.CertificateRequest:
+		return k.PublicKey, nil
+	case ssh.CryptoPublicKey:
+		return k.CryptoPublicKey(), nil
+	case *ssh.Certificate:
+		return ExtractKey(k.Key)
+	case jose.JSONWebKey:
+		return k.Key, nil
+	case *jose.JSONWebKey:
+		return k.Key, nil
+	default:
+		return nil, fmt.Errorf("cannot extract the key from type '%T'", k)
+	}
+}
+
+// VerifyPair that the public key matches the given private key.
+func VerifyPair(pubkey crypto.PublicKey, key crypto.PrivateKey) error {
+	switch pub := pubkey.(type) {
+	case *rsa.PublicKey:
+		priv, ok := key.(*rsa.PrivateKey)
+		if !ok {
+			return errors.New("private key type does not match public key type")
+		}
+		if !pub.Equal(priv.Public()) {
+			return errors.New("private key does not match public key")
+		}
+	case *ecdsa.PublicKey:
+		priv, ok := key.(*ecdsa.PrivateKey)
+		if !ok {
+			return errors.New("private key type does not match public key type")
+		}
+		if !pub.Equal(priv.Public()) {
+			return errors.New("private key does not match public key")
+		}
+	case ed25519.PublicKey:
```

Comment:
I know it's intentional, but I'm curious why ed25519's public key isn't a pointer while the rest are.

Reply:
`ed25519.PublicKey` is an alias to `[32]byte` in the Go runtime. While other types are heavier and represent a `struct`, using a pointer is preferred.